feat: expand best practices (#924)
ParsePlatform/Docs: _includes/parse-server/best-practice.md

- Use rate-limiting rules for public endpoints, for example limit the number of requests per IP address or per user.
- Use very restrictive rules for private endpoints; for example limit access to Parse Dashboard to your personal network.

> ### Security-first mindset

Reviewer: ```suggestion
### Security-First Mindset
```

> When developing Parse Server, assume that any cloud function or custom trigger is going to be abused by a malicous client. It's best to assume that the client code cannot be trusted. Even if you have written the client code, it's very simple for an attacker to replicate a request to your Parse Server and modify arguments.
> 
> The following cloud code is **not recommended** and is an example of poor security practise:

Reviewer: ```suggestion
When developing for Parse Server, consider any Cloud Function or Cloud Trigger as a potential target for a malicious attack. In general, it's best to always assume that the sender of a request cannot be trusted. Even if you have control over the client's code when shipping it, it's simple for an attacker to replicate a request to Parse Server and modify arguments, or reverse-engineer a client app and extract hard coded keys to send requests with unexpected arguments.

The following cloud code is **not recommended** and is an example of poor security practise:
```

> When developing for Parse Server, consider any Cloud Function or Cloud Trigger as a potential target for a malicious attack. In general, it's best to always assume that the sender of a request cannot be trusted. Even if you have control over the client's code when shipping it, it's simple for an attacker to replicate a request to Parse Server and modify arguments, or reverse-engineer a client app and extract hard coded keys to send requests with unexpected arguments.

Reviewer: This looks like something I've written some time ago, if I'm not mistaken. Where did you copy that from? Are we duplicating content with this?

```js
Parse.Cloud.define('updateEmail', async (req) => {
  const user = await new Parse.Query(Parse.User).get(req.params.id, { useMasterKey: true });
  if (!user) {
    throw 'This user does not exist';
  }
  user.set('email', req.params.email);
  return user.save(null, { useMasterKey: true });
});
```

With this code, an attacker can simply guess a user to sniff out a user object and replace their email, potentially exposing private user data. This could allow the attacker to then request a password reset to the new email, and take-over the account completely.

```js
while (true) {
  let id = '';
  const chars = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghiklmnopqrstuvwxyz'.split('');
  for (let i = 0; i < 10; i++) {
    id += chars[Math.floor(Math.random() * chars.length)];
  }
  try {
    const user = await Parse.Cloud.run('updateEmail', {id});
    console.log(`Here is a full user object: `, user.toJSON());
    return;
  } catch (e) {
    console.log(`Not a valid id: ${id}`);
  }
}
```

> Regardless of how secure your Parse Server is with ACLs and CLPs, a few lines of insecure cloud functions can undo it.

Reviewer: ```suggestion
Regardless of how secure your Parse Server is with ACLs and CLPs, a few lines of insecure Cloud Functions can undo it.
```

A secure solution would be to ensure that users can only update their email, such as:

```js
Parse.Cloud.define('updateEmail', async (req) => {
  req.user.set('email', req.params.email);

>   await req.user.save(null, { useMasterKey: true });

Reviewer: Wouldn't it be best practice to avoid any operation using `useMasterKey: true` but instead give only the user write permission in the object ACL? This would mitigate the risk that through a developer mistake in code someone could execute operations because they are using the master key; instead any permission would be bound to the user's session token. Using `useMasterKey: true` should only be necessary if the object is locked even for the user with read/write limited to master key.
